Evaluate results of %{%expr%} blocks in statusline

shadmansaleh opened the pull request:
Currently results of %{} blocks gets directly displayed in statusline.
Because of results of %{} blocks cann't contain expressions like
%#highlight# or %f etc . This pr changes that .

The idea is simple it just replaces the %{} block with it's result and continue evaluating.

Without the patch
![Screenshot_without](https://user-images.githubusercontent.com/13149513/117556125-6070a280-b087-11eb-944d-3913aa042db6.png)

With the patch
![Screenshot_with](https://user-images.githubusercontent.com/13149513/117556121-551d7700-b087-11eb-8dba-5ac7dbdf8c78.png)


Edit:
The idea has changed. Current plan is to evaluate %{} blocks only when expr starts with a '%' to remain backword compatable. So format will be %{%expr%}

shadmansaleh committed:
Evaluate results of %{} blocks in statusline

Currently results of %{} blocks gets directly displayed in statusline.
Because of results of %{} blocks can't contain expressions like
%#highlight# or %f etc . This pr changes that .

shadmansaleh commented:
Can someone tell me what's this suppose to mean
```
buffer.c(4487) : error C2275: 'char_u' : illegal use of this type as an expression
        c:\projects\vim\src\vim.h(324) : see declaration of 'char_u'
```

char_u is just an alias for unsigned char and used all over the codebase . Why is this error occuring ?

line 4487 is 
```c
    char_u *block_start = s;
```
where s itself is a pointer to char_u

shadmansaleh committed:
Add test

obcat commented on src/testdir/test_statusline.vim:
```diff
@@ -251,6 +251,18 @@ func Test_statusline()
 
+  "%{: evaluates enxpressions present in result of %{} expression
+  func! Inner_eval()
+    return '%n some other text'
```
What happens if this line is the next one?
```vim
    return '%{Outer_eval()}'
```

shadmansaleh replied:
Now nothing special will happen . When you asked it would've gotten stuck because of infinite recursion . But now evaluation stops when 100 level depth is reached.

obcat replied:
Oh, thanks. That's exactly what I was afraid of.

shadmansaleh committed:
Cap %{} evaluation to 100 level deep

brammool commented:
This is not backwards compatible, if the %{} result contains any percent sign it's likely to be broken.
I see two alternatives:  1) provide a function that can be used inside %{} to evaluate (might not work for highlighting).
2) provide a new % item and keep %{} as-is.

shadmansaleh commented:
> This is not backwards compatible, if the %{} result contains any percent sign it's likely to be broken.
I see two alternatives: 1) provide a function that can be used inside %{} to evaluate (might not work for highlighting).
>2) provide a new % item and keep %{} as-is.

How about `%[]` then ? I think `[]` aren't used for anything yet. Highlights are pretty important so I'd rather like to go with 2 to remain backword compatable.

Though to be honest I don't think there are many use cases where %{} evaluates to string with % . Even in that case the % can just be escaped with another %

shadmansaleh committed:
Add %[] element

shadmansaleh commented:
@brammool I've moved the functionality to new `%[expr]` item . Hopefully haven't missed anything . Let me know what you think about this :)

Edit:
By the way do you know why  appveyor CI is failing those errors aren't making any sense to me

brammool commented:
I think that would be an acceptable solution.  Let's await anybody else commenting.

shadmansaleh committed:
code cleanup

shadmansaleh committed:
Switch to %{:expr}

shadmansaleh commented:
@brammool I think I've a better idea then creating a new % item . We could check for a char at start of expr that cann't exist on valid vim expr. I've just pushed an implementation of that . I think valid vimL_expr doesn't contain ':' at begining(I could be wrong though) . So here result of `%{expr}` doesn't get evaluated while result of `%{:expr}` gets evaluated . This should be backword compatable and doesn't require a new item . If ':' is valid as first char ofa  vimL expression or you think ':' is confusing please let me know a char that can be used for this purpose :) . Also let me know what you think of this idea ...

brammool commented:
> @brammool I think I've a better idea then creating a new % item . We
> could check for a char at start of expr that cann't exist on valid vim
> expr. I've just pushed an implementation of that . I think valid
> vimL_expr doesn't contain ':' at begining(I could be wrong though) . So
> here result of `%{expr}` doesn't get evaluated while result of
> `%{:expr}` gets evaluated . This should be backword compatable and
> doesn't require a new item . If ':' is valid as first char ofa  vimL
> expression or you think ':' is confusing please let me know a char that
> can be used for this purpose :) . Also let me know what you think of
> this idea ...

Both can work.  The most relevant difference might be how the end is
found.  For %{} is stated that there cannot be a "}" inside the
expression.  For %[] that would be a "]", which is much more common.
But perhaps another way can be used to mark the end, since this
construct is new anyway.  Using %[expr%] could also work, and "%]" is
very unlikely to appear.

Using %{: expr } is a bit more cryptic.  How about %{% expr %} ?
Or even %{% expr %}% to make it symmetric.

-- 
hundred-and-one symptoms of being an internet addict:
253. You wait for a slow loading web page before going to the toilet.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

shadmansaleh commented:
>For %{} is stated that there cannot be a "}" inside the
expression.  For %[] that would be a "]", which is much more common.

I think if we are going with %{} structure I think we should leave the end marker as changing it will definatly be a breaking change to anyone using %{} .

> Using %{: expr } is a bit more cryptic.  How about %{% expr %} ?

I think if we aren't changing the end marker then we can switch to using `%{%expr}` . How about `%{!expr}` . Since there's already a `%!` evaluating item might be a familiar syntax .

Edit:
Ok `!` won't work as it's a valid expression element . I think we can finalize `%{%expr}` for evaluation and `%{expr}` retains it's regular behavior . I've already made the changes to use '%'. If you agree I think this proposal can be merged to vim . You can review the changes and let me know if you have any problem about the code .

shadmansaleh committed:
Change symbol from ':' to '%'

brammool commented:
> > For %{} is stated that there cannot be a "}" inside the
> > expression.  For %[] that would be a "]", which is much more common.
> 
> I think if we are going with %{} structure we can leave the end marker
> ad it is since it always has been '}' and has worked till now . Also
> changing the end marker will definatly be a breaking change to anyone
> using %{} .
> 
> > Using %{: expr } is a bit more cryptic.  How about %{% expr %} ?
> 
> I think if we aren't changing the end marker then we can switch to
> using `%{%expr}` . How about `%{!expr}` . Since there's already a `%!`
> evaluating item might be a familiar syntax .

Yes, %{! is better than %{:, since "!" often means "execute".

I do think that using %} for the end might work better.  Not allowing a
"}" inside the expression can be a problem, e.g. when using a
dictionary.  If we do not change it now it will be set in stone.

-- 
Did you hear about the new 3 million dollar West Virginia State Lottery?
The winner gets 3 dollars a year for a million years.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

shadmansaleh commented:
@brammool I think github doesn't send mails about edits .That's why you got an older version of the comment.

To sum it up here's where I'm at.

`!` would have been better but . `!` is not operator in vimL . So it can be at start of a valid statement . That's why we can't use that . Also we can't change the end marker as that will break all %{} statements out there .
For example if some one has stl=%{my_function()} that now will have to be changed to stl=%{my_function()%} . That entirely defeats  the purpose of being backword compatable with changes in this pr .

I think we can finalize %{%expr} for evaluation and %{expr} retains it's regular behavior . I've already made the changes to use '%'. If you agree I think this proposal can be merged to vim . You can review the changes and let me know if you have any problem about the code .

shadmansaleh commented:
> I do think that using %} for the end might work better.  Not allowing a
"}" inside the expression can be a problem, e.g. when using a
dictionary.  If we do not change it now it will be set in stone.

I do think It would have been better to have `%}` . But it's already on the stone . Do you really want to make  a breaking change about it ? If you do I don't have problem making it . To be clear we are no longer introducing a new item. `%[]` was droped now we are changing the behavior of already existing `%{}` item to extend it's abilities while remaining 100% backword compatable . Changing the end marker will break all existing uses of `%{}` item . I'm just asking for confirmation because I'm really confused about you trying to make incompatable changes :D .

jamessan commented:
> Changing the end marker will break all existing uses of %{} item

No, the suggestion is to require the new end marker (`%}`) when using the new start marker (`%{%`).  Existing users of `%{...}` won't be affected, but it allows for clearer parsing of `%{%...%}`.

shadmansaleh committed:
Change end marker to %} for %{% exprs

shadmansaleh commented:
> No, the suggestion is to require the new end marker (%}) when using the new start marker (%{%).

@jamessan thanks for clearifiing . Now that makes a lot more sence :) .

@brammool I've changed the end marker for `%{%` expressions to `%}`

So now for `%{%myfunc()%}` results of `myfunc()` will be evaluated and `%{myfunc()}` would act as before. This is what you asked for right ?

brammool commented:
> > No, the suggestion is to require the new end marker (%}) when using the new start marker (%{%).
> 
> @jamessan thanks for clearifiing . Now that makes a lot more sence :) .
> 
> @brammool I've changed the end marker for `%{%` expressions to `%}`
> 
> So now for `%{%myfunc()%}` results of `myfunc() will be evaluated .
> This is what you asked for right ? 

Right, only require %} instead of } when using the new start marker.

I wonder if we can use %{!, since using ! in front of an expression
would result in a boolean, which doesn't make sense to insert into the
status line.  So even though it's not invalid, it would never be
actually used.

I just like %{! a bit better than %{%, but both can work.

-- 
hundred-and-one symptoms of being an internet addict:
260. Co-workers have to E-mail you about the fire alarm to get
     you out of the building.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

jamessan commented:
> I wonder if we can use %{!, since using ! in front of an expression
would result in a boolean, which doesn't make sense to insert into the
status line.  So even though it's not invalid, it would never be
actually used.

There are lots of boolean pieces of information which can make sense to put in the status line.  I wouldn't count on `%{!` not being used.  It's syntactically valid, so why risk it?

nickspoons commented:
> > I wonder if we can use %{!, since using ! in front of an expression
> > would result in a boolean, which doesn't make sense to insert into the
> > status line.  So even though it's not invalid, it would never be
> > actually used.
> 
> There are lots of boolean pieces of information which can make sense to put in the status line. I wouldn't count on `%{!` not being used. It's syntactically valid, so why risk it?

I agree. And it would be nice if a ternary could be used here, in which case the first part of the expression _would_ be a bool. Warning: contrived example ahead:

     %{% !is_active ? resolveInactive() : resolveActive() %}

shadmansaleh commented:
@brammool I think we shouldn't use `!` . As jamessan
 nickspoons pointed out there can be valid usecases .

shadmansaleh committed:
Fix MSVC builds

shadmansaleh committed:
Fix appveyor CI

shadmansaleh commented:
@brammool is this ok to merge ? Or something left to be done ?

brammool commented:
Please also add a test for infinite recursion, so that it runs into
MAX_STL_EVAL_DEPTH.

Please use formatting like in the other Vim code: { on a separate line,
empty line after declarations.


-- 
       Bravely bold Sir Robin, rode forth from Camelot,
       He was not afraid to die, Oh Brave Sir Robin,
       He was not at all afraid to be killed in nasty ways
       Brave, brave, brave, brave Sir Robin.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

brammool commented:
> > I do think that using %} for the end might work better.  Not allowing a
> > "}" inside the expression can be a problem, e.g. when using a
> > dictionary.  If we do not change it now it will be set in stone.
> 
> I do think It would have been better to have `%}` . But it's already on
> the stone . Do you really want to make  a breaking change about it ? If
> you do I don't have problem making it . To be clear we are no longer
> introducing a new item. `%[]` was droped now we are changing the
> behavior of already existing `%{}` item to extend it's abilities while
> remaining 100% backword compatable . Changing the end marker will break
> all existing uses of `%{}` item . I'm just asking for confirmation
> because I'm really confused about you trying to make incompatable
> changes :D .

We can use %} only for %{%, not for the existing %{.  That gives the new
form an advantage and doesn't change the existing one.

-- 
    A KNIGHT rides into shot and hacks him to the ground.  He rides off.
    We stay for a moment on the glade.  A MIDDLE-AGED LADY in a C. & A.
    twin-set emerges from the trees and looks in horror at the body of her
    HUSBAND.
MRS HISTORIAN: FRANK!
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

shadmansaleh committed:
Fix code formating

shadmansaleh committed:
Add test for inf recursion

shadmansaleh commented:
> Please also add a test for infinite recursion, so that it runs into
MAX_STL_EVAL_DEPTH.
>
>Please use formatting like in the other Vim code: { on a separate line,
empty line after declarations

Thanks for the review. I've added test for inf recursion and fixed formating as you recomanded. I actually don't know what formating rules vim follow , didn't see anything related to formatting in CONTRIBUTING.md
